Enable xdoctest runner in CI for real this time

Erotemic opened the pull request:
Builds on #83317 and enables running the doctests. Just need to figure out what is causing the failures.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

Erotemic committed:
Re-enable doctests

Erotemic committed:
Remove unused code

Erotemic committed:
cleanup hipify

Erotemic committed:
Doctest linting

Erotemic committed:
More doctest linting

Erotemic committed:
Add requirements to cpp and profiler doctests

Erotemic committed:
linting

Erotemic committed:
lint and shift around skips

Erotemic commented:
Latest commits disable more of the doctests in hopes that some subset of them are causing the errors, if so I'll start to hone in on exactly which ones are problematic.

While doing this I started linting the doctests to help ensure they also conform with pep8 standards.

Erotemic committed:
Fix two doctests

Erotemic committed:
Attempt disabling quantized experiments

Erotemic committed:
Disable more doctest to debug

Erotemic committed:
Disable more tests

Erotemic committed:
Disable more tests

Erotemic committed:
Disable more tests

Erotemic committed:
dispatch has exposed an xdoctest bug. working around it.

Erotemic committed:
Disable more tests

Erotemic committed:
Try very small subset of tests

Erotemic committed:
Bump xdoctest version

Erotemic committed:
only test nn doctests

Erotemic commented:
So I've found a subset of tests that doesn't cause any issue. Specifically the ones in torch.utils.

That means simply executing a doctest on the CI isn't an issue. There is are specific ones that are causing a segfault and that teardown error on the CI.

Erotemic committed:
Verify errors still exist

Erotemic committed:
Test ao

Erotemic committed:
The ao module seems ok

ezyang commented:
If you want, we can enable the tasks one by one, landing each incremental PR each time. I can give you rights to use ghstack to setup the stack.

ezyang commented:
Sgtm

Erotemic committed:
Squashed test commits

ezyang commented:
I'm going to turn off notifications on this PR for now, reping me when you're ready

Erotemic committed:
Test fx doctests

Erotemic committed:
Test monitor tests

Erotemic committed:
Test futures

Erotemic committed:
Retest all

Erotemic committed:
Cleanup

Erotemic commented:
@ezyang I think the offending test was the one in torch.monitor that was invoking tensorboard. Disabling that in addition to the other tests I disabled in this PR seems to allow the other tests to run correctly.

This PR should now be stable, and new doctests will automatically be added and run in the test suite.

Erotemic commented:
This CLA should have been signed. I got approval. I'll go ahead an ping the appropriate people.

kit1980 commented:
/easycla

Erotemic commented:
@ezyang I've finally completed the CLA process. I've rebased and force pushed. Hopefully everything still works.

Erotemic committed:
Skip failing doctests

Erotemic committed:
Bump xdoctest version, extra skip for pre 1.1.0 decorator bug

Erotemic committed:
Fix doctest syntax

Erotemic commented:
@ezyang The dashboard is green, the CLA is signed. This PR is ready.

ezyang commented on torch/_functorch/eager_transforms.py:
```diff
@@ -1393,11 +1399,12 @@ def functionalize(func: Callable, *, remove: str = 'mutations') -> Callable:
 
     Example::
 
+        >>> # xdoctest: +SKIP
```
We should probably make some spreadsheet for the skips and encourage people to work on them 😝

Erotemic replied:
I agree. Fortunately they are very easy to grep for! You can also put a string explanation in them too, which I stopped doing because I've already put a ton of time into this.

In some cases the skip makes sense if the example code isn't meant to be executed, but I do think it's better when you can verify that a piece of code actually executes with some demo information.

ezyang commented on torch/autograd/function.py:
```diff
@@ -48,6 +48,7 @@ def save_for_backward(self, *tensors: torch.Tensor):
         See :ref:`extending-autograd` for more details on how to use this method.
 
         Example::
+            >>> # xdoctest: +REQUIRES(env:TORCH_DOCTEST_AUTOGRAD)
```
So how exactly do the REQUIRES annotations work, do we have to remember to put then on all the autograd related doctests?

Erotemic replied:
Short answer:

* The `+REQUIRES(env:...)` directive tests for an environment variable and skips the test if its not there.

* This is not required for new tests.

Long answer:

When xdoctests parses an example block, it breaks it into parts based on a few criteria like: "is there a got/want test?, or "is there a directive?". In the case of a directive, before it executes the code after it, it checks to see if the directive is both satisfied (based on the directive name and its arguments) and enabled (which is denoted by the `+` and nearly always true if you using directives, but technically you can have an `# xdoctest -REQUIRES(env:TORCH_DOCTEST_AUTOGRAD)`, which disables it. 

The takeaway is that if there is a `+REQUIRES(env:FOO)`, directive, that test will be skipped unless the `FOO` environment variable exists and is truthy in `os.environ`. (site node: you can also do this to test for the presense of dependencies, e.g. `+REQUIRES(module:torchvision)`).  

In `test/run_tests.py`, I have the `run_doctests` enable or disable these environment variables depending the environment the tests are running in. Currently `TORCH_DOCTEST_AUTOGRAD` is never enabled, likely because I found that the tests were causing issues. So having this environment variable serves as a quick way to enable / disable all of the associated tests, which can ideally allow them to be cleaned up and fixed in the future.

As for new autograd tests, you don't need to add this directive to them. If they run and they pass, that's great! However, if the reason I disabled them in the first place (I don't recall details, maybe I have a previous post that explains it) impacts the new test, then they will fail. This is a good thing, we want to know when doctests don't actually execute the way we thought they did, and we want to explicitly mark the iffy ones.

ezyang commented on torch/autograd/function.py:
```diff
@@ -48,6 +48,7 @@ def save_for_backward(self, *tensors: torch.Tensor):
         See :ref:`extending-autograd` for more details on how to use this method.
 
         Example::
+            >>> # xdoctest: +REQUIRES(env:TORCH_DOCTEST_AUTOGRAD)
```
So how exactly do the REQUIRES annotations work, do we have to remember to put then on all the autograd related doctests?

ezyang commented:
@pytorchbot merge

malfet commented:
@pytorchbot rebase

Erotemic commented:
It doesn't look like the current failures are caused by anything in this branch:

```
UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - py-lief==0.12.0 -> python[version='>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
```

Is that correct, or might I be including something that is causing this.

ezyang commented:
@pytorchbot rebase

ezyang commented:
I don't know. But it's also xmas so I don't want to land something risky without good evidence that everything is OK. Let's see if the rebase helps

ezyang commented:
This looks real and doctest related https://github.com/pytorch/pytorch/actions/runs/3769972417/jobs/6409635888 

I advise not running doctest on M1

ezyang commented:
The docker build failures are also legit (although not your fault). You need a green docker build everywhere to be able to push but the focal builds are failing. https://github.com/pytorch/pytorch/pull/91371 should help

Erotemic committed:
Fix doctest parse errors

Erotemic committed:
Mark doctest dependencies

Erotemic commented:
I rebased on master and force pushed. There seemed to be only 14 failing tests on the M1.

```
torch/distributed/optim/zero_redundancy_optimizer.py ZeroRedundancyOptimizer:0
torch/distributed/_composable/replicate.py replicate:0
torch/distributed/_composable/_ddp.py DistributedDataParallel.register_comm_hook:0
torch/distributed/_composable/_ddp.py DistributedDataParallel.register_comm_hook:1
torch/distributed/_shard/sharded_tensor/__init__.py custom_sharded_op_impl:0
torch/distributed/_shard/sharding_plan/api.py ShardingPlan:0
torch/testing/_internal/distributed/_tensor/common_dtensor.py skip_unless_torch_gpu:0
torch/ao/nn/quantized/functional.py conv1d:0
torch/ao/nn/quantized/functional.py conv2d:0
torch/ao/nn/quantized/functional.py conv3d:0
torch/ao/nn/quantized/modules/linear.py Linear:0
torch/ao/nn/quantized/modules/conv.py Conv1d:0
torch/ao/nn/quantized/modules/conv.py Conv2d:0
torch/ao/nn/quantized/modules/conv.py Conv3d:0
```

The first 7 of these also failed on macos-12. They seemed to be caused by 2 underlying issues. For the first 7, the `torch._C` module didn't exist, so I added a `# xdoctest: +REQUIRES(module:torch._C)`, which should conditionally skip those tests. For the second m1 only failures, they seem to be due to not having a QEngine, so I tagged those with the exiting `xdoctest: +REQUIRES(env:TORCH_DOCTEST_QENGINE)` condition, which is currently hard coded to be off in run_tests.py. Future work can set this more granularity.

But if these are the only issues, then I think we can keep doctests enabled on M1.

Erotemic committed:
doctest fixes

ezyang commented:
Ok sgtm!

Erotemic committed:
Skip more doctests

Erotemic committed:
another disabled doctest

ezyang commented:
The explanation about torch._C is sus, because I would expect this module to exist on M1 mac

Erotemic committed:
Remove REQUIRES torch._C directives

Erotemic commented:
It looks like you are correct about `torch._C`. I thought that was the issue because I got an error saying that `module:torch._C._distributed_c10d` could not be imported because `torch._C` was not a package, but the requires check for torch._C didn't fix it. However, replacing that with `# xdoctest: +REQUIRES(module:torch._C._distributed_c10d)` did seem to do the trick.

Last night when I rebased on master I think I inadvertently removed a change you made to include #91371. I rebased on master again this morning, not sure if that includes the fix for the build CI or not, but I don't see immediate errors, so that is a good sign.

ezyang commented:
I just merged my PR, so I think all we need to do is rebase past it

ezyang commented:
@pytorchbot rebase

ezyang commented:
nuuuu you pushed lol. I was pretty sure the previous ci was going to be ok. I've confirmed the lief errors are not your fault and was going to bypassed them

Erotemic commented:
It had already failed by the time I force pushed. Your POSIX branch (57dcd93c4103c6db043f341a0242596a42188081) is merged in what I recently pushed.

Either way, if you want to try something, I won't touch it.

ezyang commented:
well, now we wait. The failures I'm expecting to see are the py-lief binary ones, as well as a mypy failure on sm86

ezyang commented:
@pytorchbot merge -f "all failures are preexisting failures"
